### ISSUE TITLE
Audit example fails with 'NoneType' object has no attribute '_is_clean'

### ISSUE THREAD

**Issue by valhuber** [valhuber/ApiLogicServer#10]
The [Tutorial](https://github.com/valhuber/ApiLogicServer/wiki/Tutorial#customize-with-python) fails as noted in the title.  It also fails in FAB, with a different error (something about an additional session being created).

**Note:** you need ApiLogicServer 1.4.5 (released last night), and rebuild the demo sample.  This picks up a required change in nw.sqlite db...

To create the issue:
```
curl -X PATCH "http://localhost:5000/Employee/1/"\
  -H  "accept: application/vnd.api+json"\
  -H  "Content-Type: application/json"\
  -d "{  \"data\": {\
        \"attributes\": {\
            \"Salary\": \"2000000\"    },\
  \"type\": \"Employee\",\
  \"id\": \"1\"  }}"
```

This is possibly a Logic Bank bug, but I have run the same test without flask and it's session magic.  I will verify...

Log:
```
Logic Phase:		FLUSH   (sqlalchemy flush processing       	 - 2021-02-08 22:07:55,997 - logic_logger - DEBUG
[2021-02-08 22:07:55,998] ERROR: 'NoneType' object has no attribute '_is_clean'
Traceback (most recent call last):
  File "/Users/val/Desktop/ApiLogicServer/venv/lib/python3.8/site-packages/safrs/safrs_api.py", line 607, in method_wrapper
    result = fun(*args, **kwargs)
  File "/Users/val/Desktop/ApiLogicServer/venv/lib/python3.8/site-packages/safrs/jsonapi.py", line 317, in patch
    instance = self._patch_instance(data, id)
  File "/Users/val/Desktop/ApiLogicServer/venv/lib/python3.8/site-packages/safrs/jsonapi.py", line 353, in _patch_instance
    instance._s_patch(**attributes)
  File "/Users/val/Desktop/ApiLogicServer/venv/lib/python3.8/site-packages/safrs/base.py", line 289, in _s_patch
    safrs.DB.session.commit()
  File "/Users/val/Desktop/ApiLogicServer/venv/lib/python3.8/site-packages/sqlalchemy/orm/scoping.py", line 163, in do
    return getattr(self.registry(), name)(*args, **kwargs)
  File "/Users/val/Desktop/ApiLogicServer/venv/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 1042, in commit
    self.transaction.commit()
  File "/Users/val/Desktop/ApiLogicServer/venv/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 504, in commit
    self._prepare_impl()
  File "/Users/val/Desktop/ApiLogicServer/venv/lib/python3.8/site-packages/sqlalchemy/orm/session.py", line 481, in _prepare_impl
    if self.session._is_clean():
AttributeError: 'NoneType' object has no attribute '_is_clean'
[2021-02-08 22:07:55,999] ERROR: 'NoneType' object has no attribute '_is_clean'
```

**Comment by valhuber** [comment]
Note that the session differ from what safrs is committing, vs. what logicBank saw during before_flush
<img width="1772" alt="different session" src="https://user-images.githubusercontent.com/5579631/107366306-d5393e00-6a92-11eb-8a37-0c22e15e68c8.png">
<img width="1772" alt="different session" src="https://user-images.githubusercontent.com/5579631/107366314-d79b9800-6a92-11eb-9a0d-7722a696e796.png">

**Comment by valhuber** [comment]
Beginning to wonder if event firing is different for SQLAlchemy vs flask-SQLAlchemy.  

But, could also be logging is replicating lines..?

Here is a log from LogicBank project (not flask) doing employee update:
```
Test 2 - alter Salary, ensure EmployeeAudit created

Logic Phase:		BEFORE COMMIT(session=0x102c741f0)          						 - 2021-02-09 12:26:09,289 - logic_logger - DEBUG
Logic Phase:		ROW LOGIC(session=0x102c741f0) (sqlalchemy before_flush)			 - 2021-02-09 12:26:09,289 - logic_logger - DEBUG
..Employee[1] {Update - client} Id: 1, LastName: Davolio, FirstName: Nancy, Title: Sales Representative, TitleOfCourtesy: Ms., BirthDate: 1980-12-08, HireDate: 2024-05-01, Address: 507 - 20th Ave. E. Apt. 2A, City: don't audit this, Region: North America, PostalCode: 98122, Country: USA, HomePhone: (206) 555-9857, Extension: 5467, Photo: None, Notes: Education includes a BA in psychology from Colorado State University in 1970.  She also completed 'The Art of the Cold Call.'  Nancy is a member of Toastmasters International., ReportsTo: 2, PhotoPath: http://accweb/emmployees/davolio.bmp, IsCommissioned: 1, Salary:  [91000.00-->] 136500.000, order_count_sql:  [None-->] 0  row: 0x102e335e0  session: 0x102c741f0 - 2021-02-09 12:26:09,289 - logic_logger - DEBUG
..Employee[1] {BEGIN NW Copy to: EmployeeAudit} Id: 1, LastName: Davolio, FirstName: Nancy, Title: Sales Representative, TitleOfCourtesy: Ms., BirthDate: 1980-12-08, HireDate: 2024-05-01, Address: 507 - 20th Ave. E. Apt. 2A, City: don't audit this, Region: North America, PostalCode: 98122, Country: USA, HomePhone: (206) 555-9857, Extension: 5467, Photo: None, Notes: Education includes a BA in psychology from Colorado State University in 1970.  She also completed 'The Art of the Cold Call.'  Nancy is a member of Toastmasters International., ReportsTo: 2, PhotoPath: http://accweb/emmployees/davolio.bmp, IsCommissioned: 1, Salary:  [91000.00-->] 136500.000, order_count_sql:  [None-->] 0  row: 0x102e335e0  session: 0x102c741f0 - 2021-02-09 12:26:09,292 - logic_logger - DEBUG
....EmployeeAudit[None] {Insert - Copy EmployeeAudit} Id: None, EmployeeId: None, LastName: Davolio, FirstName: Nancy, Title: Sales Representative, Salary: 136500.000, CreatedOn: None  row: 0x102e33eb0  session: 0x102c741f0 - 2021-02-09 12:26:09,292 - logic_logger - DEBUG
....EmployeeAudit[None] {early_row_event_all_classes - handle_all sets 'Created_on} Id: None, EmployeeId: None, LastName: Davolio, FirstName: Nancy, Title: Sales Representative, Salary: 136500.000, CreatedOn: 2021-02-09 12:26:09.292795  row: 0x102e33eb0  session: 0x102c741f0 - 2021-02-09 12:26:09,292 - logic_logger - DEBUG
Logic Phase:		COMMIT(session=0x102c741f0)   										 - 2021-02-09 12:26:09,293 - logic_logger - DEBUG
..Employee[1] {Commit Event} Id: 1, LastName: Davolio, FirstName: Nancy, Title: Sales Representative, TitleOfCourtesy: Ms., BirthDate: 1980-12-08, HireDate: 2024-05-01, Address: 507 - 20th Ave. E. Apt. 2A, City: don't audit this, Region: North America, PostalCode: 98122, Country: USA, HomePhone: (206) 555-9857, Extension: 5467, Photo: None, Notes: Education includes a BA in psychology from Colorado State University in 1970.  She also completed 'The Art of the Cold Call.'  Nancy is a member of Toastmasters International., ReportsTo: 2, PhotoPath: http://accweb/emmployees/davolio.bmp, IsCommissioned: 1, Salary:  [91000.00-->] 136500.000, order_count_sql:  [None-->] 0  row: 0x102e335e0  session: 0x102c741f0 - 2021-02-09 12:26:09,293 - logic_logger - DEBUG
Logic Phase:		FLUSH(session=0x102c741f0)   (sqlalchemy flush processing)       	 - 2021-02-09 12:26:09,293 - logic_logger - DEBUG


Ran 1 test in 0.031s

OK
```

But equivalent test in ApiLogicServer (with flask):
```

Logic Phase:		BEFORE COMMIT(session=0x111d59f10)          						 - 2021-02-09 12:19:53,769 - logic_logger - DEBUG
Logic Phase:		ROW LOGIC(session=0x111d59f10) (sqlalchemy before_flush)			 - 2021-02-09 12:19:53,769 - logic_logger - DEBUG
..Employee[1] {Update - client} Id: 1, LastName: Davolio, FirstName: Nancy, Title: Sales Representative, TitleOfCourtesy: Ms., BirthDate: 1980-12-08, HireDate: 2024-05-01, Address: 507 - 20th Ave. E. Apt. 2A, City: Seattle, Region: North America, PostalCode: 98122, Country: USA, HomePhone: (206) 555-9857, Extension: 5467, Photo: None, Notes: Education includes a BA in psychology from Colorado State University in 1970.  She also completed 'The Art of the Cold Call.'  Nancy is a member of Toastmasters International., ReportsTo: 2, PhotoPath: http://accweb/emmployees/davolio.bmp, IsCommissioned: 1, Salary:  [91000.0000000000-->] 2000000  row@: 0x111d8a460  session@: 0x111d59f10 - 2021-02-09 12:19:53,769 - logic_logger - DEBUG
..Employee[1] {early_events} Id: 1, LastName: Davolio, FirstName: Nancy, Title: Sales Representative, TitleOfCourtesy: Ms., BirthDate: 1980-12-08, HireDate: 2024-05-01, Address: 507 - 20th Ave. E. Apt. 2A, City: Seattle, Region: North America, PostalCode: 98122, Country: USA, HomePhone: (206) 555-9857, Extension: 5467, Photo: None, Notes: Education includes a BA in psychology from Colorado State University in 1970.  She also completed 'The Art of the Cold Call.'  Nancy is a member of Toastmasters International., ReportsTo: 2, PhotoPath: http://accweb/emmployees/davolio.bmp, IsCommissioned: 1, Salary:  [91000.0000000000-->] 2000000  row@: 0x111d8a460  session@: 0x111d59f10 - 2021-02-09 12:19:53,770 - engine_logger - DEBUG
..Employee[1] {formula_rules} Id: 1, LastName: Davolio, FirstName: Nancy, Title: Sales Representative, TitleOfCourtesy: Ms., BirthDate: 1980-12-08, HireDate: 2024-05-01, Address: 507 - 20th Ave. E. Apt. 2A, City: Seattle, Region: North America, PostalCode: 98122, Country: USA, HomePhone: (206) 555-9857, Extension: 5467, Photo: None, Notes: Education includes a BA in psychology from Colorado State University in 1970.  She also completed 'The Art of the Cold Call.'  Nancy is a member of Toastmasters International., ReportsTo: 2, PhotoPath: http://accweb/emmployees/davolio.bmp, IsCommissioned: 1, Salary:  [91000.0000000000-->] 2000000  row@: 0x111d8a460  session@: 0x111d59f10 - 2021-02-09 12:19:53,770 - engine_logger - DEBUG
..Employee[1] {Constraint END Constraint Function: <function declare_logic.<locals>.raise_over_20_percent at 0x11177e160>  on ..Employee[1]: Id: 1, LastName: Davolio, FirstName: Nancy, Title: Sales Representative, TitleOfCourtesy: Ms., BirthDate: 1980-12-08, HireDate: 2024-05-01, Address: 507 - 20th Ave. E. Apt. 2A, City: Seattle, Region: North America, PostalCode: 98122, Country: USA, HomePhone: (206) 555-9857, Extension: 5467, Photo: None, Notes: Education includes a BA in psychology from Colorado State University in 1970.  She also completed 'The Art of the Cold Call.'  Nancy is a member of Toastmasters International., ReportsTo: 2, PhotoPath: http://accweb/emmployees/davolio.bmp, IsCommissioned: 1, Salary:  [91000.0000000000-->] 2000000  row@: 0x111d8a460  session@: 0x111d59f10} Id: 1, LastName: Davolio, FirstName: Nancy, Title: Sales Representative, TitleOfCourtesy: Ms., BirthDate: 1980-12-08, HireDate: 2024-05-01, Address: 507 - 20th Ave. E. Apt. 2A, City: Seattle, Region: North America, PostalCode: 98122, Country: USA, HomePhone: (206) 555-9857, Extension: 5467, Photo: None, Notes: Education includes a BA in psychology from Colorado State University in 1970.  She also completed 'The Art of the Cold Call.'  Nancy is a member of Toastmasters International., ReportsTo: 2, PhotoPath: http://accweb/emmployees/davolio.bmp, IsCommissioned: 1, Salary:  [91000.0000000000-->] 2000000  row@: 0x111d8a460  session@: 0x111d59f10 - 2021-02-09 12:19:53,770 - engine_logger - DEBUG
..Employee[1] {early_events} Id: 1, LastName: Davolio, FirstName: Nancy, Title: Sales Representative, TitleOfCourtesy: Ms., BirthDate: 1980-12-08, HireDate: 2024-05-01, Address: 507 - 20th Ave. E. Apt. 2A, City: Seattle, Region: North America, PostalCode: 98122, Country: USA, HomePhone: (206) 555-9857, Extension: 5467, Photo: None, Notes: Education includes a BA in psychology from Colorado State University in 1970.  She also completed 'The Art of the Cold Call.'  Nancy is a member of Toastmasters International., ReportsTo: 2, PhotoPath: http://accweb/emmployees/davolio.bmp, IsCommissioned: 1, Salary:  [91000.0000000000-->] 2000000  row@: 0x111d8a460  session@: 0x111d59f10 - 2021-02-09 12:19:53,771 - engine_logger - DEBUG
..Employee[1] {BEGIN Copy to: EmployeeAudit} Id: 1, LastName: Davolio, FirstName: Nancy, Title: Sales Representative, TitleOfCourtesy: Ms., BirthDate: 1980-12-08, HireDate: 2024-05-01, Address: 507 - 20th Ave. E. Apt. 2A, City: Seattle, Region: North America, PostalCode: 98122, Country: USA, HomePhone: (206) 555-9857, Extension: 5467, Photo: None, Notes: Education includes a BA in psychology from Colorado State University in 1970.  She also completed 'The Art of the Cold Call.'  Nancy is a member of Toastmasters International., ReportsTo: 2, PhotoPath: http://accweb/emmployees/davolio.bmp, IsCommissioned: 1, Salary:  [91000.0000000000-->] 2000000  row@: 0x111d8a460  session@: 0x111d59f10 - 2021-02-09 12:19:53,771 - logic_logger - DEBUG

Logic Phase:		BEFORE COMMIT(session=0x111d59f10)          						 - 2021-02-09 12:19:53,771 - logic_logger - DEBUG
....EmployeeAudit[None] {Insert - Copy EmployeeAudit} Id: None, Title: Sales Representative, Salary: 91000.0000000000, LastName: Davolio, FirstName: Nancy, EmployeeId: None, CreatedOn: None  row@: 0x111d80430  session@: 0x111d59f10 - 2021-02-09 12:19:53,774 - logic_logger - DEBUG
....EmployeeAudit[None] {early_row_event_all_classes - handle_all sets 'Created_on} Id: None, Title: Sales Representative, Salary: 91000.0000000000, LastName: Davolio, FirstName: Nancy, EmployeeId: None, CreatedOn: 2021-02-09 12:19:53.774272  row@: 0x111d80430  session@: 0x111d59f10 - 2021-02-09 12:19:53,774 - logic_logger - DEBUG
....EmployeeAudit[None] {early_events} Id: None, Title: Sales Representative, Salary: 91000.0000000000, LastName: Davolio, FirstName: Nancy, EmployeeId: None, CreatedOn: 2021-02-09 12:19:53.774272  row@: 0x111d80430  session@: 0x111d59f10 - 2021-02-09 12:19:53,774 - engine_logger - DEBUG
....EmployeeAudit[None] {formula_rules} Id: None, Title: Sales Representative, Salary: 91000.0000000000, LastName: Davolio, FirstName: Nancy, EmployeeId: None, CreatedOn: 2021-02-09 12:19:53.774272  row@: 0x111d80430  session@: 0x111d59f10 - 2021-02-09 12:19:53,774 - engine_logger - DEBUG
Logic Phase:		COMMIT(session=0x111d59f10)   										 - 2021-02-09 12:19:53,774 - logic_logger - DEBUG
visit: ..Employee[1]: Id: 1, LastName: Davolio, FirstName: Nancy, Title: Sales Representative, TitleOfCourtesy: Ms., BirthDate: 1980-12-08, HireDate: 2024-05-01, Address: 507 - 20th Ave. E. Apt. 2A, City: Seattle, Region: North America, PostalCode: 98122, Country: USA, HomePhone: (206) 555-9857, Extension: 5467, Photo: None, Notes: Education includes a BA in psychology from Colorado State University in 1970.  She also completed 'The Art of the Cold Call.'  Nancy is a member of Toastmasters International., ReportsTo: 2, PhotoPath: http://accweb/emmployees/davolio.bmp, IsCommissioned: 1, Salary: 91000.0000000000  row@: 0x111d8a460  session@: 0x111d59f10 - 2021-02-09 12:19:53,775 - engine_logger - DEBUG
..Employee[1] {Commit Event} Id: 1, LastName: Davolio, FirstName: Nancy, Title: Sales Representative, TitleOfCourtesy: Ms., BirthDate: 1980-12-08, HireDate: 2024-05-01, Address: 507 - 20th Ave. E. Apt. 2A, City: Seattle, Region: North America, PostalCode: 98122, Country: USA, HomePhone: (206) 555-9857, Extension: 5467, Photo: None, Notes: Education includes a BA in psychology from Colorado State University in 1970.  She also completed 'The Art of the Cold Call.'  Nancy is a member of Toastmasters International., ReportsTo: 2, PhotoPath: http://accweb/emmployees/davolio.bmp, IsCommissioned: 1, Salary: 91000.0000000000  row@: 0x111d8a460  session@: 0x111d59f10 - 2021-02-09 12:19:53,775 - logic_logger - DEBUG
visit: ....EmployeeAudit[None]: Id: None, Title: Sales Representative, Salary: 91000.0000000000, LastName: Davolio, FirstName: Nancy, EmployeeId: None, CreatedOn: 2021-02-09 12:19:53.774272  row@: 0x111d80430  session@: 0x111d59f10 - 2021-02-09 12:19:53,775 - engine_logger - DEBUG
Logic Phase:		FLUSH(session=0x111d59f10)   (sqlalchemy flush processing)       	 - 2021-02-09 12:19:53,775 - logic_logger - DEBUG
```

**Comment by valhuber** [comment]
Fixed in 4.3.2
